### PR TITLE
Add ability to retain input metadata into Pydantic models.

### PR DESCRIPTION
**Pull request by bbsbb**
## Background

This is a quick enhancement intended to be backward compatible within the current experimental pydantic model. As a future ref, this should certainly be considered for default behavior before hitting stable.

Achieving GraphQL semantics similar to "PATCH" is usually implemented with Optional input. 
The GraphQL spec suggests that missing optional input is not coerced to values, but rather stays out of the input map ( https://spec.graphql.org/October2021/#sec-Input-Objects.Input-Coercion ).

In Strawberry, a decision has been made to handle this with its own value - UNSET.

Pydantic, on the other side, long has had the ability to introspect fields not provided on instance creation( https://pydantic-docs.helpmanual.io/usage/exporting_models/ ). This is a particularly useful feature when relying on the library as a validation layer and bridge to domain objects.
Currently in Strawberry the only potential workaround to retain the information of "What fields were provided within the request on input?" requires:

* dropping reliance on strawberry.auto and explicitly swapping factories while also introducing incoherent typing
or
* leaking strawberry abstractions within the domain models.

## Description

The proposed solution introduces a backward compatible way to retain the information of a value's presence at input up to the Pydantic validator.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

Decided not to open an issue and use this instead.

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Boyan Bonev for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: tests/experimental/pydantic/test_basic.py
Comment (thejaminator) [inline tests/experimental/pydantic/test_basic.py:71]: can we add another testcase / statement with
```
friend: Optional[str] = None 
```

default value with None is tricky in pydantic context...

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [inline strawberry/experimental/pydantic/utils.py:93]: i think theres a way where we don't need this behaviour in get_default_factory_for_field. let me try.

File: tests/experimental/pydantic/test_basic.py
Comment (bbsbb) [reply tests/experimental/pydantic/test_basic.py:71]: Done, 4070d5f.
Let me know if another change is desired.

File: strawberry/experimental/pydantic/utils.py
Comment (bbsbb) [reply strawberry/experimental/pydantic/utils.py:93]: Indeed. It can be inlined in type annotation and to_pydantic. I can take a swing of that as well(or another implementation) if you would provide a bit more context ;o)

Btw, I was wondering if there is an RFC or other type of document providing some background on the current concept of strawberry/pydantic interop. Would be interesting to learn more about what are the conceptual edges and how was the current implementation born, because I have a feeling that instead of interfacing with, strawberry is trying to *be* pydantic, which, to me, was suprising(and I would have said undesired), but I expect there are concerns I am missing out of lack of knowledge.

File: strawberry/experimental/pydantic/object_type.py
Comment (BryceBeagle) [inline strawberry/experimental/pydantic/object_type.py:202]: Minor formatting improvement:
```suggestion
            instance_kwargs = dataclasses.asdict(
                self,
                dict_factory=(
                    dict if not with_input_metadata else PydanticInputMetadataDict
                ),
            )
```

File: docs/integrations/pydantic.md
Comment (BryceBeagle) [inline docs/integrations/pydantic.md:272]: Minor styling adjustment:
```suggestion
When looking to use pydantic as a validation layer and export models afterwards,
it is sometimes important to retain metadata about the original input. In Strawberry,
this is possible via the `with_input_metadata` flag on input models.
```

Comment (BryceBeagle) [reply docs/integrations/pydantic.md:272]: I think this part needs to be rephrased though:
> When looking to use pydantic as a validation layer and export models afterwards

Perhaps
> When using pydantic as a validation layer, it is sometimes important to retain metadata about the original input while exporting models.

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: docs/integrations/pydantic.md
Comment (bbsbb) [reply docs/integrations/pydantic.md]: Change reflected and pushed. Thank you for improving the wording!

File: strawberry/experimental/pydantic/utils.py
Comment (thejaminator) [reply strawberry/experimental/pydantic/utils.py:95]:
```
def test_default_strawberry_type_with_input_metadata():
    class User2(pydantic.BaseModel):
        friend: Optional[str] = pydantic.Field(default_factory=lambda: "friend_value")
        frenemy: str

    @strawberry.experimental.pydantic.type(User2, with_input_metadata=True)
    class UserType2:
        friend: strawberry.auto
        frenemy: strawberry.auto

    user_type_2 = UserType2(frenemy="asd")
    assert user_type_2.friend == "friend_value"
```

so if you run this, the test will fail. because of the short-circuit for `get_default_factory_for_field` :( 
you'll see that in `test_default_and_default_factory`.
 the default from pydantic should be retrieved.  (well, i'm not the maintainer of this repo, so thats my opnion)

so instead of the short-circuit in `get_default_factory_for_field`, maybe retrieve the unset fields in a different function instead?

regarding the RFC, i didn't see any for it, maybe @patrick91 can comment

File: tests/experimental/pydantic/test_basic.py
Comment (thejaminator) [inline tests/experimental/pydantic/test_basic.py:107]: hmmm if you try this you can see that if you pass the default explictly, the value from the passed value doesn't get used

```
    # test when we pass the default value
    user_input_default = UserType2(
        frenemy="frenemy_value", friend="not_default_friend"
    )
    assert user_input_default.to_pydantic().friend == "not_default_friend"
    assert user_input_2.to_pydantic().dict() == {
        "friend": "not_default_friend",
        "frenemy": "frenemy_value",
    }
    assert user_input_2.to_pydantic().dict(exclude_unset=True) == {
        "friend": "not_default_friend",
        "frenemy": "frenemy_value",
    }
```

File: tests/experimental/pydantic/test_basic.py
Comment (thejaminator) [inline tests/experimental/pydantic/test_basic.py:79]: add tests for  defaults converted from pydantic
```suggestion
    assert user_input_1.friend == "friend_value"
    assert user_input_1.to_pydantic().friend == "friend_value"
```

File: strawberry/experimental/pydantic/utils.py
Comment (patrick91) [reply strawberry/experimental/pydantic/utils.py:95]: > regarding the RFC, i didn't see any for it, maybe @patrick91 can comment

Unfortunately there's no RFC, the Pydantic integration has always been an experiment :D it definitely needs some polishing

I wonder if this is something we can discuss at the next strawberry monthly meeting, I don't use Pydantic a lot at the moment so I do lack knowledge around it unfortunately

File: strawberry/experimental/pydantic/object_type.py
Comment (BryceBeagle) [reply strawberry/experimental/pydantic/object_type.py:43]: (You'll need to add an import for `Iterable`)

Comment (bbsbb) [reply strawberry/experimental/pydantic/object_type.py]: Added.

File: tests/experimental/pydantic/test_basic.py
Comment (bbsbb) [reply tests/experimental/pydantic/test_basic.py:107]: Hey @thejaminator  - Just want to make sure here - you are creating an input called user_input_default and then testing user_input_2 (which is the one created at line 97). I believe there may have been a copy paste mix up? Hit me up on discord if we need to discuss something!

Comment (thejaminator) [reply tests/experimental/pydantic/test_basic.py:108]: sure, i replied to you what i want to test for in the discord chat https://discord.com/channels/689806334337482765/928742670447304734/930456003915444224

**Review by BryceBeagle — APPROVED** [review]

File: docs/integrations/pydantic.md
Comment (patrick91) [inline docs/integrations/pydantic.md:296]: can we add some explaination on why this is useful? 😊

Comment (bbsbb) [reply docs/integrations/pydantic.md:296]: Will address this and some comments from discord over the weekend!

Comment (patrick91) [reply docs/integrations/pydantic.md:296]: @bbsbb will you be able to update the docs before I merge this? 😊

**Comment by kidig** [comment]
@bbsbb May I ask why you closed it without merging?

**Pull request closed without merging**
